tests/models/test_study.py

```diff
@@ -0,0 +1,482 @@
+
+    with pytest.raises(exc.IntegrityError, match=r"study_rule_name_idx"):
+        db_session.add(studyrule.StudyRule(name="dupe"))
+        db_session.commit()
+
+
+def test_study_rule__defaults_created_date(create_study_rule_db, db_session):
+    db_session.add(studyrule.StudyRule(name="asdf"))
+    db_session.commit()
+    s = db_session.query(studyrule.StudyRule).filter(studyrule.StudyRule.name == "asdf").first()
+    assert s.created_date is not None
+
+
+def test_study_rule__defaults_updated_date(create_study_rule_db, db_session):
+    db_session.add(studyrule.StudyRule(name="asdf"))
+    db_session.commit()
+    s = db_session.query(studyrule.StudyRule).filter(studyrule.StudyRule.name == "asdf").first()
+    assert s.updated_date is not None
+
+
+def test_study_rule__updates_updated_date(create_study_rule_db, db_session):
+    db_session.add(studyrule.StudyRule(id=1618, name="asdf"))
+    db_session.commit()
+
+    db_session.query(studyrule.StudyRule).filter(studyrule.StudyRule.id == 1618).update({"name": "updated"})
+    db_session.commit()
+
+    updated = db_session.query(studyrule.StudyRule).filter(studyrule.StudyRule.id == 1618).first()
+    assert updated.updated_date > updated.created_date
+
+
+def test_study_rule__whole_programs(create_study_rule_db, db_session):
+    """Test that study.whole_programs populate correctly.
+
+    Each study should contain it's corresponding records from
+    StudyRuleProgram.
+    """
+    db_session.add(studyrule.StudyRule(id=1618, name="first"))
+    db_session.add(studyrule.StudyRule(id=3141, name="second"))
+    db_session.commit()
+
+    db_session.add(studyrule.StudyRuleProgram(study_rule_id=1618, program_name="alpha"))
+    db_session.add(studyrule.StudyRuleProgram(study_rule_id=1618, program_name="beta"))
+    db_session.commit()
+
+    first_study = db_session.query(studyrule.StudyRule).filter(studyrule.StudyRule.id == 1618).first()
+    assert len(first_study.whole_programs) == 2
+    assert all([program.study_rule_id == 1618 for program in first_study.whole_programs])
+
+    second_study = db_session.query(studyrule.StudyRule).filter(studyrule.StudyRule.id == 3141).first()
+    assert len(second_study.whole_programs) == 0
+
+
+def test_study_rule__partial_programs(create_study_rule_db, db_session):
+    """Test that study.partial_programs populate correctly.
+
+    Each study should contain it's corresponding records from
+    StudyRuleProgramProject.
+    """
+    db_session.add(studyrule.StudyRule(id=1618, name="first"))
+    db_session.add(studyrule.StudyRule(id=3141, name="second"))
+    db_session.commit()
+
+    db_session.add(studyrule.StudyRuleProgramProject(study_rule_id=1618, program_name="alpha", project_code="uno"))
+    db_session.add(studyrule.StudyRuleProgramProject(study_rule_id=1618, program_name="alpha", project_code="dos"))
+    db_session.commit()
+
+    first_study = db_session.query(studyrule.StudyRule).filter(studyrule.StudyRule.id == 1618).first()
```

Comment:
It doesn't really make a difference because there are other tests to check for uniqueness, but in the context of a test, you can assert that there's only one matching row as a side effect by using `one()` rather than `first()`.

Reply:
didn't know about `one()`.  i like it better as it enforces the intent.